### PR TITLE
Fix data race and deadlock in file_integrity module

### PR DESCRIPTION
**Pull request by adriansr**
This PR fixes two issues with auditbeat's file_integrity module:
- Concurrent map write in #8009 
Auditbeat's file_integrity module had a data race in its recursive watcher implementation for Linux. Was caused by file watches being installed from different goroutines.
This patch fixes the problem by moving watch installation to the same goroutine that processes inotify's events, guaranteeing that watches are installed in order, which is necessary for consistence.

- deadlock when file integrity monitor is started
    A deadlock is possible in auditbeat's file_integrity module under Windows: When enough events arrive while watches are being installed, the event channel can fill causing the installation of a watch to block. This patch makes sure that events are received while watches are being installed, and at the same time ensures that no event is lost.

### CODE REVIEW

**Comment by adriansr** [comment]
There seems to be a problem I didn't catch in local testing

File: auditbeat/module/file_integrity/eventreader_fsnotify.go [hunk 1]
```diff
@@ -56,7 +55,15 @@ func (r *reader) Start(done <-chan struct{}) (<-chan Event, error) {
...
 	}
-	go r.consumeEvents(done)
+
+	queueDone := make(chan struct{})
+	queueC := make(chan []*Event)
+
+	// Launch a separate goroutine to fetch all events that happen while
+	// watches are being installed.
+	go func() {
+		queueC <- r.enqueueEvents(queueDone)
```
Comment (andrewkroh) [inline auditbeat/module/file_integrity/eventreader_fsnotify.go]: When the writer is done with `queueC` it should be closed.

File: auditbeat/module/file_integrity/eventreader_test.go [hunk 2]
```diff
@@ -233,6 +235,74 @@ func TestEventReader(t *testing.T) {
...
+	r, err := NewEventReader(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	done := make(chan struct{})
+	defer close(done)
+
+	// Generate a lot of events in parallel to Start() so there is a chance of
+	// events arriving before all watched dirs are Add()-ed
```
Comment (andrewkroh) [inline auditbeat/module/file_integrity/eventreader_test.go]: Good idea.

File: auditbeat/module/file_integrity/eventreader_fsnotify.go (hunk 1 above)
Comment (adriansr) [reply auditbeat/module/file_integrity/eventreader_fsnotify.go]: done

**Comment by adriansr** [comment]
@cwurm, that's how I implemented it at first. However, it introduces an inconsistency that will trigger failures in tests, as the test may be performing an action right after `Start` but before a watch is installed, causing an expected event to be lost.

**Comment by adriansr** [comment]
Yes @cwurm, that would have fixed the problem too. I felt it made more sense to confine all the accesses in the same place.

**Pull request merged**